src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,463 @@
+        idx = torch.stack(values).argsort(descending=True)
+        counts = Counter([keys[i] for i in idx[:rank_budget]])
+        counts = {k: counts.get(k, 0) for k in hook_layer_map.keys()} # add layers with 0 rank
+        for k, k_hook in equal_inputs_map.items():
+            # ensure hook layers have the highest rank if they are equal to another layer
+            rank, rank_hook = counts[k], counts[k_hook]
+            if rank_hook >= rank:
+                continue
+            counts[k_hook], counts[k] = rank, rank_hook
+        return counts
+
+    # set function to check if modules should be added to hooks
+    if hasattr(model, "peft_config"):
+        _check_fn = lambda name, module: isinstance(module, Linear)
+    else:
+        _check_fn = lambda name, module: any([name.endswith(t) for t in peft_config.target_modules])
+
+    training = model.training
+    device = get_device_with_meta_params(model)
+    model.eval()
+
+    hooks = {}
+    for name, module in model.named_modules():
+        # currently only linear layers are supported
+        if not _check_fn(name, module):
+            continue
+        hook = HashHook(name)
+        module.register_forward_hook(hook)
+        hooks[name] = hook
+    rank_budget = peft_config.r * len(hooks)
+    max_components = round(peft_config.r * peft_config.eva_config.rho)
+
+    # forward for one batch to check which layer inputs are equal to avoid unneeded svd calculations
+    inputs = next(iter(dataloader))
+    forward_fn(model, move_inputs_to_device(inputs, device))
+    hash_dict = {k: h.hashed_inputs[0] for k, h in hooks.items()}
+    equal_inputs_map = {vv: v[0] for v in find_equal_values(hash_dict).values() for vv in v[1:]}
+    
+    # initialize svd hooks
+    for k in list(hooks.keys()):
+        module = model.get_submodule(k)
+        module._forward_hooks.clear()
+        _ = hooks.pop(k)
+        if k in equal_inputs_map:
+            continue
+        if isinstance(prepare_activations_fn, Mapping):
+            try:
+                fn = prepare_activations_fn[k]
+            except KeyError:
+                raise ValueError(f"if prepare_activations_fn is a mapping it must contain adapter module name {k}")
+        else:
+            fn = prepare_activations_fn
+        hooks[k] = SVDHook(k, max_components, peft_config.eva_config.tau, fn)
+    layer_hook_map = {**dict(zip(hooks.keys(), hooks.keys())), **equal_inputs_map}
+
+    # start svd calculation
+    pbar = tqdm(iter(cycle(dataloader)), position=0, leave=False)
+    convergence_dict = {k: False for k in hooks.keys()}
+    rank_dist = {k: max_components for k in layer_hook_map.keys()}
+    for inputs in pbar:
+
+        indices = None
+        if get_indices_fn is not None:
+            indices = get_indices_fn(inputs, peft_config)
+
+        for name, hook in hooks.items():
+            module = model.get_submodule(name)
+            module._forward_hooks.clear()
```

Comment:
Hmm, should we call `hook.remove()` instead? What if there are other, unrelated forward hooks on the model?